README.md
[![Build Status](https://travis-ci.org/moznion/Perl-Lint.svg?branch=master)](https://travis-ci.org/moznion/Perl-Lint) [![Coverage Status](https://img.shields.io/coveralls/moznion/Perl-Lint/master.svg?style=flat)](https://coveralls.io/r/moznion/Perl-Lint?branch=master)
# NAME

Perl::Lint - Yet Another Perl Source Code Linter

# SYNOPSIS

    use Perl::Lint;

    my $linter = Perl::Lint->new;
    my $target_files = [qw(foo/bar.pl buz.pm)];
    my $violations   = $linter->lint($target_files);

# DESCRIPTION

Perl::Lint is the yet another source code linter for perl.

# AIMS

Development of this module aims to create a fast and flexible static analyzer for Perl5 that has compatibility with Perl::Critic

Please see also [http://news.perlfoundation.org/2014/03/grant-proposal-perllint---yet.html](http://news.perlfoundation.org/2014/03/grant-proposal-perllint---yet.html).

# METHODS

- `$linter->lint($target_files:SCALAR or ARRAYREF, $args:HASHREF)`

    `lint` checks the violations of target files. It can export.
    On default, this function checks the all of policies that are in `Perl::Lint::Policy::*`.

# PERFORMANCE

Benchmark script: [https://github.com/moznion/Perl-Lint/blob/master/author/benchmark\_lint\_vs\_critic.pl](https://github.com/moznion/Perl-Lint/blob/master/author/benchmark_lint_vs_critic.pl).

                   Rate Perl::Critic   Perl::Lint
    Perl::Critic 20.6/s           --         -78%
    Perl::Lint   92.0/s         348%           --

# SEE ALSO

[Perl::Critic](https://metacpan.org/pod/Perl::Critic)

# LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

moznion <moznion@gmail.com>


Changes
Revision history for Perl extension Perl-Lint

{{$NEXT}}

0.25 2016-09-27T02:08:05Z

    - Fix interaction with B::Keywords properly (Tom Hukins++)

0.24 2016-05-22T16:37:08Z

    - Test fixes for B::Keywords 1.15 (Tom Hukins++)

0.23 2015-05-30T04:03:15Z

    - Fix bug about analysis of dereference (@daleevans++)

0.22 2015-04-04T16:53:59Z

    - Add testing module (Test::Perl::Lint)

0.21 2015-04-03T05:26:40Z

    - Fix line number to point out the violation

0.20 2015-03-28T15:52:01Z

    - Fix some bugs
    - Implement some policies
    - Support `## no lint` annotation!

0.11 2015-01-26T16:01:45Z

    - Implement some remained policies
    - Bug fix

0.10 2014-10-24T07:21:09Z

    - original not development version

0.01_02 2014-08-29T04:48:45Z

    - Implement new policies
    - Add filter mechanism

0.01_01 2014-08-12T06:53:40Z

    - original version



lib/Perl/Lint.pm
package Perl::Lint;
use 5.010001;
use strict;
use warnings;
use Carp ();
use Compiler::Lexer;
use Module::Pluggable;
use Module::Load;

our $VERSION = "0.25";

sub new {
    my ($class, $args) = @_;

    my @ignores;

    if (my $ignores = $args->{ignore}) {
        if (ref $ignores ne 'ARRAY') {
            Carp::croak "`ignore` must be array reference";
        }

        push @ignores, map {"Perl::Lint::Policy::$_"} @$ignores;
    }

    if (my $filters = $args->{filter}) {
        if (ref $filters ne 'ARRAY') {
            Carp::croak "`filter` must be array reference";
        }

        for my $filter (@$filters) {
            my $filter_package = "Perl::Lint::Filter::$filter";
            load $filter_package;

            push @ignores, map {"Perl::Lint::Policy::$_"} @{$filter_package->filter};
        }
    }

    Module::Pluggable->import(
        search_path => 'Perl::Lint::Policy',
        require     => 1,
        inner       => 0,
        except      => [@ignores],
    );
    my @site_policies = plugins(); # Exported by Module::Pluggable

    # TODO add mechanism to add extend policies

    bless {
        args => $args,
        site_policies => \@site_policies,
    }, $class;
}

sub lint {
    my ($self, $files) = @_;

    my @files = ($files); # when scalar value
    if (my $ref = ref $files) {
        if ($ref ne 'ARRAY') {
            Carp::croak("Argument of files expects scalar value or array reference");
        }
        @files = @$files;
    }

    my @violations;
    for my $file (@files) {
        open my $fh, '<', $file or die "Cannnot open $file: $!";
        my $src = do { local $/; <$fh> };

        push @violations, @{$self->_lint($src, $file)};
    }

    return \@violations;
}

sub lint_string {
    my ($self, $src) = @_;

    return $self->_lint($src);
}

sub _lint {
    my ($self, $src, $file) = @_;

    if (!defined $src || $src eq '') {
        return [];
    }

    my $args = $self->{args};

    my $lexer = Compiler::Lexer->new($file);
    my $tokens = $lexer->tokenize($src);

    # list `no lint` lines
    # TODO improve performance
    my %no_lint_lines = ();
    my %used_no_lint_lines = ();
    my $line_num = 1;
    for my $line (split /\r?\n/, $src) {
        $line =~ s/"(?:[^"]*?#[^"]*?)*"//g;
        $line =~ s/'(?:[^']*?#[^']*?)*'//g;
        $line =~ s/\[(?:[^\[]*?#[^\]]*?)*\]//g;
        if ($line =~ /(#.+)?##\s*no\slint(?:\s+(?:qw)?([(][^)]*[)]|"[^"]*"|'[^']*'))?/) {
            next if $1; # Already commented out at before

            my $annotations = {};
            if ($2) {
                my $annotation = substr $2, 1, -1;
                $annotations = {map {$_ => 1} grep {$_} split /[, ]/, $annotation};
            }
            $no_lint_lines{$line_num} = $annotations;
        }
        $line_num++;
    }

    my $prohibit_useless_no_lint_policy;

    my @violations;
    for my $policy (@{$self->{site_policies}}) {
        if ($policy eq 'Perl::Lint::Policy::Miscellanea::ProhibitUselessNoLint') {
            $prohibit_useless_no_lint_policy = $policy;
            next;
        }

        if ($policy eq 'Perl::Lint::Policy::Miscellanea::ProhibitUnrestrictedNoLint') {
            push @violations, @{$policy->evaluate($file, $tokens, $src, $args, \%no_lint_lines)};
            next;
        }


        for my $violation (@{$policy->evaluate($file, $tokens, $src, $args)}) {
            my $violation_line = $violation->{line};
            my $no_lint = $no_lint_lines{$violation_line};
            if (!$no_lint || (keys %$no_lint > 0 && !$no_lint->{(split /::/, $violation->{policy})[-1]})) {
                push @violations, $violation;
            }
            $used_no_lint_lines{$violation_line} = 1;
        }
    }

    if ($prohibit_useless_no_lint_policy) {
        push @violations,
            @{$prohibit_useless_no_lint_policy->evaluate($file, \%no_lint_lines, \%used_no_lint_lines)};
    }

    return \@violations;
}

1;

__END__

=encoding utf-8

=head1 NAME

Perl::Lint - Yet Another Perl Source Code Linter

=head1 SYNOPSIS

    use Perl::Lint;

    my $linter = Perl::Lint->new;
    my $target_files = [qw(foo/bar.pl buz.pm)];
    my $violations   = $linter->lint($target_files);

=head1 DESCRIPTION

Perl::Lint is the yet another source code linter for perl.

=head1 AIMS

Development of this module aims to create a fast and flexible static analyzer for Perl5 that has compatibility with Perl::Critic

Please see also L<http://news.perlfoundation.org/2014/03/grant-proposal-perllint---yet.html>.

=head1 METHODS

=over 4

=item * C<< $linter->lint($target_files:SCALAR or ARRAYREF, $args:HASHREF) >>

C<lint> checks the violations of target files. It can export.
On default, this function checks the all of policies that are in C<Perl::Lint::Policy::*>.

=back

=head1 PERFORMANCE

Benchmark script: L<https://github.com/moznion/Perl-Lint/blob/master/author/benchmark_lint_vs_critic.pl>.

                   Rate Perl::Critic   Perl::Lint
    Perl::Critic 20.6/s           --         -78%
    Perl::Lint   92.0/s         348%           --

=head1 SEE ALSO

L<Perl::Critic>

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut

